src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,761 @@
+            : this(EqualityComparer<TKey>.Default, 0)
+        {
+            _backup = dict;
+        }
+
+        public SmallCapacityDictionary(IEqualityComparer<TKey> comparer)
+            : this(comparer, 0)
+        {
+        }
+
+        public SmallCapacityDictionary(int capacity)
+            : this(EqualityComparer<TKey>.Default, capacity)
+        {
+        }
+
+        public SmallCapacityDictionary(IEqualityComparer<TKey> comparer, int capacity)
+        {
+            if (comparer is not null && comparer != EqualityComparer<TKey>.Default) // first check for null to avoid forcing default comparer instantiation unnecessarily
+            {
+                _comparer = comparer;
+            }
+            else
+            {
+                _comparer = EqualityComparer<TKey>.Default;
+            }
+
+            if (capacity == 0)
+            {
+                _arrayStorage = Array.Empty<KeyValuePair<TKey, TValue>>();
+            }
+            else
+            {
+                _arrayStorage = new KeyValuePair<TKey, TValue>[capacity];
+            }
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SmallCapacityDictionary{TKey, TValue}"/> initialized with the specified <paramref name="values"/>.
+        /// </summary>
+        /// <param name="values">An object to initialize the dictionary. The value can be of type
+        /// <see cref="IDictionary{TKey, TValue}"/> or <see cref="IReadOnlyDictionary{TKey, TValue}"/>
+        /// or an object with public properties as key-value pairs.
+        /// </param>
+        /// <remarks>
+        /// If the value is a dictionary or other <see cref="IEnumerable{T}"/> of <see cref="KeyValuePair{String, Object}"/>,
+        /// then its entries are copied. Otherwise the object is interpreted as a set of key-value pairs where the
+        /// property names are keys, and property values are the values, and copied into the dictionary.
+        /// Only public instance non-index properties are considered.
+        /// </remarks>
+        public SmallCapacityDictionary(IEnumerable<KeyValuePair<TKey, TValue>> values)
+        {
+            _comparer = EqualityComparer<TKey>.Default;
+
+            if (values is IEnumerable<KeyValuePair<TKey, TValue>> keyValueEnumerable)
+            {
+                _arrayStorage = Array.Empty<KeyValuePair<TKey, TValue>>();
+
+                foreach (var kvp in keyValueEnumerable)
+                {
+                    Add(kvp.Key, kvp.Value);
+                }
+
+                return;
+            }
+
+            _arrayStorage = Array.Empty<KeyValuePair<TKey, TValue>>();
+        }
+
+        /// <inheritdoc />
+        public TValue this[TKey key]
+        {
+            get
+            {
+                if (key == null)
+                {
+                    ThrowArgumentNullExceptionForKey();
+                }
+
+                TryGetValue(key, out var value);
+
+                if (value == null)
```

Comment:
Does this collection not allow storing null values? I'd have expected:
```suggestion
                if (!TryGetValue(key, out var value))
```

Reply:
was a mistake.